### ISSUE TITLE
API design issue: "claims" and service classes

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#17324]
With #17197 (multiple schedulers), I see the third instance (load-balancers and storage being the others) of an API pattern and I want to pin it down.

The pattern is that of a "claim" being filed (pod, ingress, pvclaim) that is to be satisfied by some semi-abstract "service class", which is implemented by a process or pod somewhere in the cluster.

For load-balancing this manifests as "what kind of load balancer will you get", and for storage it manifests as "what kind of persistent volume will you get".  In both cases we have an open question as to whether the varying classes of service are reflected anywhere in the API or whether they simply understood to exist in the abstract.  Now that we have a third example (what scheduler will I get), I'd like to iron this out.

API:
- PRO: allows users to "go shopping"
- PRO: allows the system to verify that a requested class of service is valid (producing an error instead of pending)
- CON: more API objects
- CON: more complex (are they namespaced or global?  How to vary QoS by namespace?  What happens if a class is removed while claims are live?  How do we indicate that a class is "healthy"?)

Abstract:
- PRO: simpler
- PRO: scoping of policy (namespaced or not) are decided by the implementor
- CON: typo'ed claims get stuck pending without a clear reason
- CON: informal, harder to comprehend "current state"

@bgrant0607 @smarterclayton @lavalamp @markturansky @saad-ali @bprashanth

**Comment by markturansky** [comment]
Is there a reason the API server can't also have the same config (for storage, whatever) and use it to validate incoming requests?  If not, that removes one API Pro and one Abstract Con.  

Also, can I assume file watchers can reload config data dynamically and not require an API to be so dynamic?

Any good admin will document their work and the cluster's configuration.  QoS classes should be easily communicated within an organization.  I'm not sure "going shopping" is a strong enough argument.

Versioned config schemas vs. API types.  They seem roughly comparable in terms of number of objects and complexity.  If the object isn't in one, it'll be in the other.

**Comment by bprashanth** [comment]
> The pattern is that of a "claim" being filed (pod, ingress, pvclaim) that is to be satisfied by some semi-abstract "service class", which is implemented by a process or pod somewhere in the cluster.

There's a similar pattern for provisioning external resources in general (creating a DNS redcord, waiting for CA to sign a cert, acquiring a static IP etc). 
* A users asks for X by creating a claim, waits till a controller has populated the status field, then references the claim to make use of X
* A user has X, self-populates the field and creates claim, proceeds to use it

Such claims cannot typically be satisfied by someting in kube-controller-manager because the external api might throw errors for reasons unrelated to Kubernetes, and the acquisition time is too long (in fact it might even involve a human).

> For load-balancing this manifests as "what kind of load balancer will you get",

I can make an argument for composable claims too. 

I think it would be really really nice to have everything one needs to run a website done for you in 1 resource (it would have to wait on a DNS claim, and then an IP claim etc), which might be a claim of type website?

> CON: more API objects

I think we want to allow people to achieve everything with 1 top level resource (good inlining/defaulting) while also allowing the freedom to request, say, DNS record and SSL cert but reuse IP. For this we'd need tiny, composable claim objects.

**Comment by bgrant0607** [comment]
Related: #3585, #14475

**Comment by bgrant0607** [comment]
And: https://github.com/docker/docker/issues/11187

**Comment by bgrant0607** [comment]
I don't know whether a single approach would be best for all cases, but I find the Abstract approach appealing for cases where we want extensibility and/or want to allow the operator to be able to apply policy, as in #17097. Typos could be handled by creating a policy object with a list of valid classes, which I've also proposed we do for labels #15390.

**Comment by thockin** [comment]
@pmorie is producing a more concise writeup of what the storage folks have come up with in this vein.

**Comment by sputnik13** [comment]
I see that Ingress is very much an implementation of this idea where a user posts an Ingress (a claim) and the Ingress Controller asynchronously delivers on the claim.  While there are definite PROs to this, how can disparity in capabilities between cloud providers be communicated to the user?  For instance AWS and GCE have different capabilities in Load Balancers.  Specifically, GCE supports path based routing and virtual hosting while AWS does not.  This effectively means the current L7 Ingress is broken for AWS.

Other differences between cloud providers as far as Ingress is concerned below...

## L7 support in cloud providers

|           | HTTP | HTTPS | Path based | Virtual host | SNI |
| --------- | ---- | ----- | ---------- | ------------ | --- |
| AWS       |   x  |   x   |            |              |     |
| GCE       |   x  |   x   |      x     |       x      |  x  |
| Mesos     |   ?  |   ?   |      ?     |       ?      |  ?  |
| Openstack |   x  |   x   |      x     |       x      |  x  |
| Rackspace |      |       |            |              |     |
| vSphere   |   x  |   x   |      x     |       x      |  x  |

## L4 Protocol Support in Cloud Providers

|           | TCP | UDP | TLS |
| --------- | --- | --- | --- |
| AWS       |  x  |     |  x  |
| GCE       |  x  |  x  |     |
| Mesos     |  ?  |  ?  |  ?  |
| Openstack |  x  |  x  |  x  |
| Rackspace |  x  |  x  |     |
| vSphere   |  x  |  x  |  x  |

While some of the details could be a little off (I don't think they are but please let me know if there's any error) I feel there's enough differences that there will be cases where users submit claims and those claims will not be honored due to underlying differences.  In such cases users likely will want feedback sooner than later.  When an API request is directly tied to the fulfillment of the requested thing/service/resource then failure can be communicated synchronously, or the API can also do some sort of validation.  However, with a decoupling of the claim and fulfillment should there be some method by which the API can discover what API requests can or cannot be fulfilled by whatever services are deployed in the cluster?

For example, there are claim and resources, and there will be controllers or what not that fulfill or provide those requests.  When such controllers are running, should they be required to register capabilities to the API so the API can provide quick feedback to the user when a request comes in that just cannot be satisfied?

**Comment by sputnik13** [comment]
It seems like the Ingress Resource is an implementation of this idea where a user posts an Ingress (a claim) and the Ingress Controller asynchronously delivers on the claim.  While there are definite PROs to this, how can disparity in capabilities between cloud providers be communicated to the user?  For instance AWS and GCE have different capabilities in Load Balancers.  Specifically, GCE supports path based routing and virtual hosting while AWS does not.  This effectively means the current L7 Ingress is broken for AWS.

While it would be great if each API thing could be general or abstract enough as to apply to all cloud providers, when there are differences in available capabilities, is it really best to make those capabilities unavailable to all users by limiting the API to just the common subset of capabilities?  I believe some have suggested using annotations/labels to denote additional capabilities, but that puts the requirement on the user to figure out which capabilities are available for the k8s deployment they are using with no feedback from the API as to the availability of such features.

As a user if I make requests that cannot be fulfilled I'd like to be told that it cannot be fulfilled by the API.  I believe this can be done if there was a concept of a service class publishing/notifying the API of available services/capabilities or if the API could discover such things.

**Comment by sputnik13** [comment]
@pmorie do you have a draft of the writeup @thockin mentioned available somewhere?

**Comment by bgrant0607** [comment]
Also related: #31571

**Comment by bgrant0607** [comment]
Aborted draft writeup by thockin: #14095

**Comment by dims** [comment]
@thockin @bgrant0607 do we want to do something here?
